src/Adapters/Auth/AuthAdapter.js

```diff
@@ -1,20 +1,102 @@
+   * @param {Object} authData The client provided authData
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
+   * @returns {(Promise<undefined|void>|void|undefined)} resolves or returns if the applicationId is valid
+   */
+  validateAppId(appIds, authData, options, request) {
+    return Promise.resolve({});
+  }
+
+  /**
+   * Legacy usage, if provided it will be triggered when authData related to this provider is touched (signup/update/login)
+   * otherwise you should implement validateSetup, validateLogin and validateUpdate
+   * @param {Object} authData The client provided authData
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
+   * @returns {Promise<ParseAuthResponse|void|undefined>}
+   */
+  validateAuthData(authData, options, request, config) {
+    return Promise.resolve({});
+  }
+
+  /**
+   * Triggered when user provide for the first time this auth provider
+   * could be a register or the user adding a new auth service
+   * @param {Object} authData The client provided authData
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
+   * @returns {Promise<ParseAuthResponse|void|undefined>}
+   */
+  validateSetUp(authData, options, req, user) {
+    return Promise.resolve({});
+  }
+
+  /**
+   * Triggered when user provide authData related to this provider
+   * he is not logged in and has already set this provider before
+   * @param {Object} authData The client provided authData
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
+   * @returns {Promise<ParseAuthResponse|void|undefined>}
+   */
+  validateLogin(authData, options, req, user) {
+    return Promise.resolve({});
+  }
+
+  /**
+   * Triggered when user provide authData related to this provider
+   * he is logged in and has already set this provider before
+   * @param {Object} authData The client provided authData
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
+   * @returns {Promise<ParseAuthResponse|void|undefined>}
    */
-  validateAppId(appIds, authData, options) {
+  validateUpdate(authData, options, req, user) {
     return Promise.resolve({});
   }
 
-  /*
-  @param authData: the client provided authData
-  @param options: additional options
+  /**
+   * Triggered in pre authentication process if needed (like webauthn, SMS OTP)
+   * @param {Object} challengeData Data provided by the client
+   * @param {(Object|undefined)} authData Auth data provided by the client, can be used for validation
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
```

Comment:
> pre-authentication

- end lines with dot.
- use backticks for code (vars, methods)
- missing param descriptions

Reply:
```suggestion
   * @param {Object} config Parse Server config object
```